File: saas/control_app/src/app/functions/shared/stacks/usage_plan_stack.py
import typing
from aws_cdk import Stack
from constructs import Construct
import aws_cdk.aws_apigateway as apigateway


class UsagePlanStack(Stack):

    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # required
        api_id = scope.node.get_context("api_id")
        tenant_id = scope.node.get_context("tenant_id")
        stage_name = scope.node.get_context("stage_name")
        api_key_value = scope.node.get_context("api_key_value")

        # optional
        limit = scope.node.try_get_context("limit") or "10000"
        period = scope.node.try_get_context("period") or "DAY"
        rate_limit = scope.node.try_get_context("rate_limit") or "10"
        burst_limit = scope.node.try_get_context("burst_limit") or "5"

        api = typing.cast(apigateway.RestApi, apigateway.RestApi.from_rest_api_id(
            self, id="Api", rest_api_id=api_id))

        throttle = apigateway.ThrottleSettings(
            burst_limit=int(burst_limit), rate_limit=int(rate_limit))
        quota = apigateway.QuotaSettings(
            limit=int(limit), period=apigateway.Period[period])

        plan = api.add_usage_plan(f'UsagePlan-{tenant_id}',
                                  throttle=throttle, quota=quota)
        cfnPlan = typing.cast(apigateway.CfnUsagePlan, plan.node.default_child)

        cfnPlan.add_property_override("ApiStages", [
            {
                "ApiId": api_id,
                "Stage": stage_name,
                # Can also specify limits for each endpoint
                # "Throttle": {
                #     "/": {
                #         "BurstLimit": 123,
                #         "RateLimit": 123
                #     }
                # }
            }
        ])

        key = apigateway.ApiKey(
            self, 'ApiKey', api_key_name='Tenant-key', value=api_key_value)
        cfnKey = typing.cast(apigateway.CfnApiKey, key.node.default_child)
        cfnKey.add_property_override('StageKeys', [
            {
                "RestApiId": api_id,
                "StageName": stage_name
            }
        ])

        plan.add_api_key(key)

File: saas/control_app/src/TenantPipeline/lib/usage_plan_stack.ts
import * as cdk from 'aws-cdk-lib';
import * as apigw from 'aws-cdk-lib/aws-apigateway';
import { ApiGateway } from 'aws-cdk-lib/aws-events-targets';
import { Construct } from 'constructs';

// import * as sqs from 'aws-cdk-lib/aws-sqs';

export class UsagePlanStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);


    const api_id = scope.node.getContext("api_id")
    const tenant_id = scope.node.getContext("tenant_id")
    const stage_name = scope.node.getContext("stage_name")
    const api_key_value = scope.node.getContext("api_key_value")

    // optional
    const limit: number = +scope.node.tryGetContext("limit") ?? "10000"
    const period_key = (scope.node.tryGetContext("period") ?? "DAY") as keyof typeof apigw.Period
    const period = apigw.Period[period_key]
    const rate_limit: number = +scope.node.tryGetContext("rate_limit") ?? "10"
    const burst_limit: number = +scope.node.tryGetContext("burst_limit") ?? "5"

    const api =
      apigw.RestApi.fromRestApiId(this, "Api", api_id) as apigw.RestApi



    const plan = api.addUsagePlan(
      `UsagePlan-${tenant_id}`, { throttle: { rateLimit: rate_limit, burstLimit: burst_limit }, quota: { limit: limit, period: period } }
    )

    const cfnPlan = plan.node.defaultChild as apigw.CfnUsagePlan

    cfnPlan.addPropertyOverride(
      "ApiStages",
      [
        {
          "ApiId": api_id,
          "Stage": stage_name,
          // Can also specify limits for each endpoint
          // "Throttle": {
          //  "/": {
          //"BurstLimit": 123,
          //"RateLimit": 123
          // }
        }

      ],
    )

    const key = new apigw.ApiKey(
      this, "ApiKey", { apiKeyName: "Tenant-key", value: api_key_value }
    )
    const cfnKey = key.node.defaultChild as apigw.CfnApiKey
    cfnKey.addPropertyOverride(
      "StageKeys", [{ "RestApiId": api_id, "StageName": stage_name }]
    )

    plan.addApiKey(key)
  }
}
